### PR TITLE
Support tensor.to("spyre", max=...) for max-strided HBM reservation

### LINKED ISSUE #2434: Runtime support for max-sized HBM and stride derivation for symbolic dims
### What feature would you like to be added?

We're enabling symbolic shape support in torch-spyre, starting with a symbolic first dimension. The initial front-end compiler side changes are in flight (#2284, #2287, #2288, #2289), but the runtime needs a few hooks before symbolic kernels can actually run.  This issue is the discussion ticket for the corresponding runtime APIs, before we lock the interface surface.

The Deeptools team has shared the contract their backend assumes for symbolic-dim tensors:

***Let's assume your tensor has shape [d0=64, d1=symbolic, d2=8] with d1 granularity=16 and max=1024. If you want to run with an actual d1=32, that is a legal value, and the program will automatically adapt to only process 32 elements in the d1 dimension, however in terms of how the data must be laid out in memory, d0 elements are stored 1 apart, d1 elements are stored 64 apart (1*d0), d2 elements are stored 65536 apart (1*d0*d1max). These strides are constant, not depending on the actual d1 size at runtime. Because of this, in practice the size of the tensor in memory will be equal to the max all the time. The amount of data we have is still only [64,32,8], and only these elements will be read/processed/written by the program, but it needs to be stored sparsely in memory following the strides above***

Two consequences:

- HBM must be sized at max along the symbolic dim.
- Device strides must be derived from max (not the per-call runtime value) so they stay constant across invocations.

The runtime today sizes HBM and computes device strides from the warmup-time shape. That's fine for static compilation, but for a symbolic kernel:

- HBM is too small the moment a call uses a value larger than the warmup.
- Device strides encode the per-call runtime N instead of N_max, so they're wrong on every non-warmup call.

Closing this gap is a prerequisite for symbolic kernels to execute correctly.

### Why is this feature needed?

Without it, symbolic kernels produced by the frontend compiler cannot execute correctly, as the runtime would size HBM and emit strides against the input tensor shape, which breaks Deeptools ' requirement for symbolic input. 

### Any alternative solution that you may want to share?

Dynamic compilation runs after `.to("spyre")`, so neither side knows `max` at allocation time. This needs further discussion on implimentation strategies. 

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by nethra-khandige**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Adds `tensor.to("spyre", max=N)`, letting the buffer be reserved at the runtime ceiling instead of the shape in front of .to() at call time, through several key modifications:

Reservation state on the tensor (`spyre_tensor_impl.h`, `spyre_tensor_impl.cpp`):

- Adds `std::optional<int64_t> reserved_dim`, `reserved_max` to `SpyreTensorImpl`
- Propagates both through `shallow_copy_and_detach_core/shallow_copy_from` so the reservation survives views/.detach(), not just the exact tensor object .to() returned
-
Allocation and resize (`spyre_mem.h`, `spyre_mem.cpp`):

- Introduces `spyre_empty_reserved`(size, stride, dtype, dim, max_size): pads size[dim] to max_size before computing SpyreTensorLayout/storage size, while the tensor's logical size/stride stay the real, unpadded values
- Extends `spyre_resize_` to pin the layout at reserved_max for a reserved tensor regardless of the requested concrete size, allow growing or shrinking within [*, max] to reuse the existing storage, and reject ever falling through to the reallocate path

Python entry point (`module.cpp`, `_monkey_patch.py`, `_C.pyi`):

- Binds `spyre_empty_reserved` to Python; adds a max= branch to spyre_to() that validates the transfer is CPU → spyre only, ensures the runtime is initialized, then calls spyre_empty_reserved + copy_tensor

Tests (`tests/tensor/test_max_reserved.py`):

- Covers shape/data correctness, the no-realloc contract (grow/shrink/round-trip within the reservation, checked via both data_ptr() and untyped_storage().nbytes()), every input-validation edge case, and reservation survival across .detach()

The enhancement lets users call tensor.to("spyre", max=N) so a compiled kernel can absorb any runtime batch size in [min, max] without reallocating the buffer or recompiling.

#### Which issue(s) this PR is related to:
Needed for symbolic shape support.

To enable: #43 
Fixes [#2434](https://github.com/torch-spyre/torch-spyre/issues/2434)
#### Does this PR introduce a user-facing change?

Yes. tensor.to("spyre", max=N) is a new, opt-in kwarg on .to() for CPU → spyre transfers; existing .to("spyre") call sites are unaffected.

### CODE REVIEW

**Comment by nethra-khandige** [comment]
Test file:

```
import torch
import torch._dynamo as dynamo
import torch.nn.functional as F

DEVICE = torch.device("spyre")
torch.manual_seed(0xAFFE)

def gelu_fn(a):
    return F.gelu(a)

x = torch.rand(560, 1024, dtype=torch.float16)
x_device = x.to(DEVICE, max = 616)
dynamo.mark_dynamic(x_device, 0, min=56, max=616)
compiled_fn = torch.compile(gelu_fn)
compiled_result = compiled_fn(x_device).cpu()
```

command to run and test: `TORCH_SPYRE_DEBUG=1 python3 test_file.py`
